pydantic_ai_slim/pydantic_ai/models/__init__.py

```diff
@@ -506,15 +509,49 @@ def cached_async_http_client(*, provider: str | None = None, timeout: int = 600,
 @cache
 def _cached_async_http_client(provider: str | None, timeout: int = 600, connect: int = 5) -> httpx.AsyncClient:
     return httpx.AsyncClient(
-        transport=_cached_async_http_transport(),
+        transport=_get_transport_for_loop(),
         timeout=httpx.Timeout(timeout=timeout, connect=connect),
         headers={'User-Agent': get_user_agent()},
     )
 
 
 @cache
-def _cached_async_http_transport() -> httpx.AsyncHTTPTransport:
-    return httpx.AsyncHTTPTransport()
+def _get_transport_for_loop() -> _PerLoopTransport:
+    return _PerLoopTransport()
+
+
+# transport._pool._connections[0]._connection._network_stream._stream.transport_stream._transport._loop
+class _PerLoopTransport(httpx.AsyncBaseTransport):
+    def __init__(self):
+        self.transports: MutableMapping[asyncio.AbstractEventLoop, httpx.AsyncHTTPTransport] = (
+            weakref.WeakKeyDictionary()
+        )
```

Comment:
It might be debatable if we want this to be a WeakKeyDictionary or a normal dict.

It would probably work with a normal dict, but it feels that if a loop is referenced by nobody but this mapping, we really don't need to have it here, so I'd say it makes sense to keep it.

Note that if we found a way to remove the strong reference to the loop, e.g. by closing the pool, then we would probably be happy with the WeakKeyDictionary without the for loop in `get_transport`, but I'm not sure I know when we should close the pool for this to stay useful (in the end, the goal is to reuse the tcp connections in the end)

Like they say, 2 hardest problems in compsci: naming things and **cache invalidation** and off-by-one errors.